src/twisted/internet/defer.py

```diff
@@ -855,6 +856,53 @@ def uncancel(result):
         return self
 
 
+    @classmethod
+    def fromCoroutine(cls, coro):
+        """
+        Schedule the execution of a coroutine that awaits/yields from
+        L{Deferred}s, wrapping it in a L{Deferred} that will fire on
+        success/failure of the coroutine.
+
+        Coroutine functions return a coroutine object, similar to how
+        generators work. This function turns that coroutine into a Deferred,
+        meaning that it can be used in regular Twisted code. For example::
+
+            import treq
+            from twisted.internet.defer import ensureDeferred
+            from twisted.internet.task import react
+
+            async def crawl(pages):
+                results = {}
+                for page in pages:
+                    results[page] = await treq.content(await treq.get(page))
+                return results
+
+            def main(reactor):
+                pages = [
+                    "http://localhost:8080"
+                ]
+                d = ensureDeferred(crawl(pages))
+                d.addCallback(print)
+                return d
+
+            react(main)
+
+        @since: Twisted NEXT
+
+        @param coro: The coroutine object to schedule.
+        @type coro: A Python 3.5+ C{async def} coroutine or a Python 3.4+
+            C{yield from} coroutine.
+
+        @raise ValueError: If C{coro} is not a coroutine or generator.
+
+        @rtype: L{Deferred}
+        """
+        if not iscoroutine(coro) and not isinstance(coro, types.GeneratorType):
+            raise ValueError("%r is not a coroutine" % (coro,))
```

Comment:
If you’re going to use a built-in exception here, it should really be TypeError since this is an error about types not values. But it should be a specific subclass of TypeError particularly given that it’s caught below; NotACoroutineError or something; maybe inherit from TypeError if you want to be able to classify it that way.

Reply:
+1 to subclass of `TypeError`.

Reply:
NotACoroutineError added, subclass of TypeError